tracer/src/Datadog.Trace.ClrProfiler.Native/method_rewriter.cpp

```diff
@@ -0,0 +1,559 @@
+#include "method_rewriter.h"
+#include "cor_profiler.h"
+#include "il_rewriter_wrapper.h"
+#include "logger.h"
+#include "stats.h"
+#include "version.h"
+#include "environment_variables_util.h"
+
+namespace trace
+{
+
+/// <summary>
+/// Rewrite the target method body with the calltarget implementation. (This is function is triggered by the ReJIT
+/// handler) Resulting code structure:
+///
+/// - Add locals for TReturn (if non-void method), CallTargetState, CallTargetReturn/CallTargetReturn<TReturn>,
+/// Exception
+/// - Initialize locals
+///
+/// try
+/// {
+///   try
+///   {
+///     try
+///     {
+///       - Invoke BeginMethod with object instance (or null if static method) and original method arguments
+///       - Store result into CallTargetState local
+///     }
+///     catch
+///     {
+///       - Invoke LogException(Exception)
+///     }
+///
+///     - Execute original method instructions
+///       * All RET instructions are replaced with a LEAVE_S. If non-void method, the value on the stack is first stored
+///       in the TReturn local.
+///   }
+///   catch (Exception)
+///   {
+///     - Store exception into Exception local
+///     - throw
+///   }
+/// }
+/// finally
+/// {
+///   try
+///   {
+///     - Invoke EndMethod with object instance (or null if static method), TReturn local (if non-void method),
+///     CallTargetState local, and Exception local
+///     - Store result into CallTargetReturn/CallTargetReturn<TReturn> local
+///     - If non-void method, store CallTargetReturn<TReturn>.GetReturnValue() into TReturn local
+///   }
+///   catch
+///   {
+///     - Invoke LogException(Exception)
+///   }
+/// }
+///
+/// - If non-void method, load TReturn local
+/// - RET
+/// </summary>
+/// <param name="moduleHandler">Module ReJIT handler representation</param>
+/// <param name="methodHandler">Method ReJIT handler representation</param>
+/// <returns>Result of the rewriting</returns>
+HRESULT TracerMethodRewriter::Rewrite(RejitHandlerModule* moduleHandler, RejitHandlerModuleMethod* methodHandler)
+{
+    auto tracerMethodHandler = static_cast<TracerRejitHandlerModuleMethod*>(methodHandler);
+
+    if (tracerMethodHandler->GetIntegrationDefinition() == nullptr)
+    {
+        Logger::Warn(
+            "NotifyReJITCompilationStarted: IntegrationDefinition is missing for "
+            "MethodDef: ", 
+            methodHandler->GetMethodDef());
+
+        return S_FALSE;
+    }
+
+    auto _ = trace::Stats::Instance()->CallTargetRewriterCallbackMeasure();
```

Comment:
silly nit: let's put this line at the begining of the method :)

Reply:
The reason it's placed a few lines below the start of the method is because the assertion that precedes it (the `GetIntegrationDefinition() == nullptr`) was used to be inside `RejitHandler::NotifyReJITParameters`, right before entering into the rewriting. As the `Tracer` and the `Debugger` need different assertions, I had to move the specific ones to the rewriters (`TracerMethodRewriter` and `DebuggerMethodRewriter`).
To keep the same order of assertions as it used to be, I placed the measuring statement immediately after the final assertion.
Does that make sense? and if it does, do you still think I should move it to the start of the method for clarity?

Reply:
We looked at the benchmarks together and haven't seen performance degradation (neither CPU nor Memory).
As for the assertion, I'll reconsider placing it at the beginning of the method as a separate PR, as part of the slydog.
Resolving.